### ISSUE TITLE
onAfterUpload function being passed an undefined file parameter.

### ISSUE THREAD

**Issue by NCDiesel** [veliovgroup/Meteor-Files#783]
I'm having an issue:

Using this example:

[GridFS Bucket Integration](https://github.com/veliovgroup/Meteor-Files/blob/e8b397df283425bfbc1d115da19878ca0d5962ac/docs/gridfs-bucket-integration.md)

I created the tooling necessary to upload files created on my server to Mongo GridFS, bucket style.    My code is very nearly 100% identical to the code in that document.  The only difference is there addition of some extra logging and validity checks and the addition of "proceedAfterUpload = true".

However, occasionally there is a hiccup in the onAfterUpload handler.   About 40-60% of the time the "file" parameter passed to the onAfterUpload function is undefined.  The bucket parameter is correct and valid, just "file" is undefined.   Again, it does not happen all the time.  Couple notes:

- The collection record is created successfully.  In other words the record for GridFSCollectionName is created and looks perfect.   Obviously GridFSCollectionName.files and and .chunks have no records related to the file being uploaded since the upload handler fails 
- The very first upload always work and I've tested and restarted this app probably a hundred times.   The first upload always works.  Coincidence maybe but I thought it was odd and noticeable.   After that whether or not it works is hit and miss.
- The file being uploaded exists on the server and is valid and has correct permissions(the files is created by the server in fact).
- This problem spans changes in versions (occurred from meteor 1.10 on and at least one prior version of ostrio:files)

Versions:

- ostrio-files 1.14.0
- meteor 1.12.1

OS:  Linux(red hat)

Issue is server side (no client interaction).

Logs:  Against corporate policy to post them but I did turn on debugging and the log indicates Meteor-files thinks the file uploaded (ie got [FilesCollection][addfile(<filename>)] with no subsequent messages.

**Comment by jankapunkt** [comment]
Would it bei possible to provide the logging and validation code so I can reproduce this issue?

If it's not possible, could you provide some code that acts similar, with the same return values etc.?

**Comment by NCDiesel** [comment]
So for the same reasons I couldn't share the logs I can't share the code.  I was being nearly 100% literal when I say I followed the example in the document I referenced in my first post.  The only thing not mentioned in the document is a call addFile().   But my use of that is unremarkable.    I do use the callback but only capture the err parameter(but addFile never errors out).   Tomorrow I will capture the fileRef parameter but I suspect it will be fine since the record gets added to the initial FSCollection just fine.  My validation for the uploadHandler is a simple.  At the beginning of the function I added:

`if (!file) {
  console.log(file)
  console.error(util.inspect(bucket);
  console.trace()
}`

The file = undefined and the bucket looks normal and valid.  The console.trace() indicates line 1435 in server.js is the culprit that doesn't pass the fileRef:

`        this.collection.insert(result, (insertErr, _id) => {
          if (insertErr) {
            callback && callback(insertErr);
            this._debug(`[FilesCollection] [addFile] [insert] Error: ${result.name} -> ${this.collectionName}`, insertErr);
          } else {
            const fileRef = this.collection.findOne(_id);
            callback && callback(null, fileRef);
            if (proceedAfterUpload === true) {
              this.onAfterUpload && this.onAfterUpload.call(this, fileRef);
              this.emit('afterUpload', fileRef); <<<<<<<<====1435
`

Also, right before I call FSCollection.addFile() I added:

`const statsResults = fs.syncstat(file);
console.error(util.inspect(statsResults));`

to confirm the file's size, perms, etc.  The file exists, is correct, and is valid.  I suspect a timing thing since whether or not it works seems fairly random.

I hope this answers your questions

**Comment by jankapunkt** [comment]
Thank you. This is a shot in the blue, but could you try to async await the `fs.stat` instead of using `fs.statSync`:

```javascript
const util = require('util')
const fsStat = util.promisify(fs.stat)

// ... other code
// ... in onAfterUpload

const statsResults = await fsStat(file)
console.error(util.inspect(statsResults))
```

the fs sync methods usually mess up the event loop so maybe this is why the emit function is not working properly.

**Comment by NCDiesel** [comment]
Sorry, I should have been more explicit - the statSync was part of the validation and troubleshooting I was doing.  I just recently added it to rule out that possibly the file being corrupt, still open, etc were the issues.   The problem existed before I added statSync. 

Important note:    This morning I added the fileRef to my addFile() call and when the upload fails the fileRef in the callback is null also.   So:

    myFSCollection.addFile(params, parms, (err, fileRef) => {
      if (err) {
        console.log('error when adding ', err);   <<<<===. This is never executed - even when it fails
      } else {
        console.dir(fileRef)     <<<<<===== This is always executed.   When it fails fileRef is undefined
      }
    });


Thanks for hanging in there and help Jan, I appreciate it.   Any continued ideas will be appreciated.

**Comment by dr-dimitru** [comment]
@jankapunkt @NCDiesel [this one looks related](https://github.com/veliovgroup/Meteor-Files/issues/779#issuecomment-742858416). Now I'm wonder when and why it was introduced 🤔 It can be in Meteor, Node.js mongodb-native-driver, or mongo server itself.

@NCDiesel tell us more about your MongoDB setup.

1. Do you have ReplicaSet in place?
    2. If so, how many members/arbiters in your set?
3. Do you utilize OpLog?
    4. If so, is it native or Redis-powered?

**Comment by NCDiesel** [comment]
Thanks for the link - I'll study it.   We do use a replicaSet and there are 3 members.  We do not use OpLog - we did until recently but the DBA killed it for performance reasons.

Currently I am adding my own async/await, checking for the existence of, and retrieving, the FSCollection record before running the upload handler manually(setting proceedAfterUpload=false).   Will let you know how that goes.

**Comment by dr-dimitru** [comment]
@NCDiesel I'm available for hire, and can help you to resolve it after signing NDA ;)
From this point I'm pretty sure it's related to MongoDB setup, OpLog configuration (*there are ways to set it up properly without trade off on performance/resources*), and parameters passed into "connection string". For example if we are writing to master and read from closest member. Or writing without waiting for propagation to entire ReplicaSet.

Forgot to ask:

5. MongoDB server version;

**Comment by NCDiesel** [comment]
Thanks.   Mongo server is 4.4 I think.

I got it working by bypassing the proceedAfterUpload, then basically creating my own upload handler and calling it like so

      addFile(....)
      // function with setTimeout(every 2**i) loop for a total of 6 reps(about 1 minute)
      // and a FSCollection.findOne()
      const gotIt = await waitForFSCollectionInsertToFinish(id). 
      if (gotIt) {
        myCustomUpLoadHandler(gotIt)
      }

So basically bypassing the event listener model and just control timing myself.  The DB is managed by another group.

**Comment by dr-dimitru** [comment]
@NCDiesel patching workaround 🐒
This issue with Mongo may arise in other parts of the app. Hope it won't be a case 🤞

**Comment by dr-dimitru** [comment]
I assume this once is solved
Feel free to reopen it in case if the issue is still persists on your end.
